### PR TITLE
add new bool func

### PR DESCRIPTION
**Pull request by yincongcyincong**
fix vm issue: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6637
however, i have to add key word bool , then the pr work, is it an expected result? it's very confused.

### CODE REVIEW

**Comment by yincongcyincong** [comment]
> fix vm issue: [VictoriaMetrics/VictoriaMetrics#6637](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6637) however, i have to add key word bool , then the pr work, is it an expected result? it's very confused.

@jiekun could you please check it, thanks.

**Comment by yincongcyincong** [comment]
just i said, the pr need key word **bool** to work
image

**Comment by yincongcyincong** [comment]
this case result in a wrong result. 
image
but this case result in a right result.
image

**Comment by jiekun** [comment]
Thank you for the new PR. I will take a look tomorrow and provide feedback in code aspect. approve from valyala and f41gh7 is needed to add this feature.

File: binaryop/funcs.go [hunk 1]
```diff
@@ -107,3 +107,19 @@ func Ifnot(left, right float64) float64 {
 	}
 	return nan
 }
+
+// And returns 1 if left and right is more than 0. Otherwise 0 is returned.
+func And(left, right float64) float64 {
+	if left > 0 && right > 0 {
```
Comment (useredsa) [inline binaryop/funcs.go:113]: The sign of the numbers should not affect the and.

Comment (useredsa) [reply binaryop/funcs.go:113]: Also, for an and without the modifier `bool`, the returned value is left.

Comment (yincongcyincong) [reply binaryop/funcs.go:113]: > Also, for an and without the modifier `bool`, the returned value is left.

then 0 and 1 should return 0，it's a little bit weird

Comment (yincongcyincong) [reply binaryop/funcs.go]: > The sign of the numbers should not affect the and.

now usigned number return 1, thanks bro.

**Comment by jiekun** [comment]
> however, i have to add key word bool , then the pr work, is it an expected result? 

You could debug it with single step. And it's should be easy to notice that the result of `(0 > 1)` is `NaN`, not `0`:
```go
func TestAndOp(t *testing.T) {
	mql := "1 and (0 > 1)"
	expr, err := Parse(mql)
	if err != nil {
		t.Fatalf("unexpected error: %s", err)
	}
	fmt.Println(expr)
}
```

```go
		evalCmp := func(cf func(left, right float64) bool) float64 {
                        // `isBool` is false
			if isBool {
				if cf(left, right) {
					return 1
				}
				return 0
			}
			if cf(left, right) {
				return left
			}
                        // It goes here and return `math.NaN`
			return nan
		}
```

So `1 and (0 > 1)` is equal to `1 and NaN`. It returns `1` in current implementation:
```
func And(left, right float64) float64 {
	if left != 0 && right != 0 {
		return left
	}
	return nan
}
```

**Comment by jiekun** [comment]
It's possible that `AND` and `OR` operation should be implemented like comparison binaryOp. 

I would like to add extra [commit](https://github.com/VictoriaMetrics/metricsql/pull/31/commits/b6a354e4fa12f5cccaf8224445757992fed74c70) for you.

I don't know the background for not implementing `AND` and `OR` operation so it's up to @valyala and @f41gh7 's review. Feel free to revert my commit if it's not the expected result.

**Comment by jiekun** [comment]
Hi @yincongcyincong I think this PR might be hold for now. As explained in the original [issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6637), `AND` is working as logical/set binary operator.

Apologize for my lack of knowledge.

**Comment by yincongcyincong** [comment]
> Hi @yincongcyincong I think this PR might be hold for now. As explained in the original [issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6637), `AND` is working as logical/set binary operator.
> 
> Apologize for my lack of knowledge.

so i think it should return an error.

**Comment by jiekun** [comment]
> so i think it should return an error.

I believe so. I think it's great to test and compare with PromQL.

**Comment by useredsa** [comment]
Why should it return an error?

The fact that is a logical / set operation does not mean this should not work this way, right? In MetricsQL 0 and 1 are oficially vectors, there are no scalars. So, the fact that it's a logical / set operation should only imply that group_left / group_right cannot be used and that the matching semantics are changed. `1 and NaN` should not give you anything, and it is like this if you label the series with `alias` or `label_set` instead of using them as is. Please refer to the original issue.

**Comment by jiekun** [comment]
@yincongcyincong I apologize for the misleading comments and reviews that have wasted your valuable time and efforts. I am currently working to address this issue.

**Pull request closed without merging**